k8scrdclient/k8scrdclient.go

```diff
@@ -86,18 +86,34 @@ func (c *CRDClient) EnsureDeleted(ctx context.Context, crd *apiextensionsv1beta1
 }
 
 func (c *CRDClient) ensureCreated(ctx context.Context, crd *apiextensionsv1beta1.CustomResourceDefinition) error {
-	c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating CRD %#q", crd.Name))
-
-	_, err := c.k8sExtClient.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
-	if errors.IsAlreadyExists(err) {
-		// Fall through. We need to check CRD status.
-		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("not creating CRD %#q", crd.Name))
-		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("CRD %#q already created", crd.Name))
-	} else if err != nil {
-		return microerror.Mask(err)
+	// We explicitly check for the existence of the CRD by fetching it to
+	// workaround nasty side effects with error handling of multiple creation
+	// attempts. It may happen that a CRD creation is forbidden while it exists
+	// and actually has to be updated.
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding CRD %#q", crd.Name))
+
+		_, err := c.k8sExtClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.Name, metav1.GetOptions{})
```

Comment:
Don't know if this is an issue here, but a _get or create_ pattern has a potential race condition here if there's a chance of this running concurrently which I'm assuming to be the case based on the comment. I.e. the window is small but if this client executes the get while another is executing create, the get may return _not found_ while the other write is being persisted.

Would it work to try the get in an else block instead? I.e. try create, if already exists fall through as before, else if err is forbidden then try get and see if found, and so on.

Reply:
The reason why I do not want to do the create is because the error handling is broken in the undrlying k8s client IMO. We see that when you call create with a CRD that exists, there are cases where the call to create returns a "forbidden" error, instead of an "already exists" error. I wanted to work around that issue. Therefore the explicit get. You are right about the race condition. I think it is just not a real problem for us. In case that happens we retry if in doubt and then it is fine again. This should then also only happen in case a completely new installation is setup where actually no CRD exists. All other cases are fine.